### PR TITLE
fix: handle non-function and malformed tools defensively

### LINKED ISSUE #32: [BUG] TypeError: Cannot read properties of undefined (reading 'parameters') when tools contain non-function tools
### Problem Description / 问题描述

In `src/stream/tool-schema.ts`, both `slimOpenAIToolsForCursor` and `buildMcpToolDefinitions` assume that every element in the `tools` array is a standard function tool containing a valid `.function` property:

```typescript
export function slimOpenAIToolsForCursor(tools: OpenAIToolDef[]): OpenAIToolDef[] {
  if (!isSlimToolsEnabled()) return tools;
  return tools.map((tool) => {
    const fn = tool.function;
    const parameters =
      fn.parameters && typeof fn.parameters === "object"
        ? (slimJsonSchema(fn.parameters) as Record<string, unknown>)
        : fn.parameters;
...
```

When an upstream extension, middleware, or client injects non-function tools into the request (for example, server-side builtin tools like `{ type: "web_search" }`, or custom tools without a `function` wrapper), `fn` evaluates to `undefined`. This immediately triggers an unhandled `TypeError`:

```
TypeError: Cannot read properties of undefined (reading 'parameters')
    at slimOpenAIToolsForCursor (src/stream/tool-schema.ts:134:10)
    at buildMcpToolDefinitions (src/stream/tool-schema.ts:158:20)
    at handleCursorNativeRequest (src/stream/native-core.ts:634:15)
    at streamSimple (src/stream/pi-adapter.ts:380:12)
```

The error aborts the turn before any request can be dispatched to Cursor.

Terminal error trace and code location

---

### Expected Behavior / 预期行为

The tool preparation pipeline should be defensive against non-function or malformed tool definitions:
1. `slimOpenAIToolsForCursor`: If a tool definition does not possess a `.function` object (or has a non-object `tool`), it should either be passed through unmodified or skipped gracefully.
2. `buildMcpToolDefinitions`: Non-function tools (or tools without a valid `.function.name`) cannot be represented as Cursor MCP client tools anyway, so they should be filtered out rather than crashing the process.

---

### Proposed Fix / 修复方案

- In `slimOpenAIToolsForCursor`:
  ```typescript
  return tools.map((tool) => {
    const fn = tool?.function;
    if (!fn) return tool;
    ...
  });
  ```
- In `buildMcpToolDefinitions`:
  ```typescript
  const result = prepared
    .filter((tool) => Boolean(tool?.function?.name))
    .map((tool) => { ... });
  ```

A PR with regression tests will be submitted shortly.

### PR DESCRIPTION
**Pull request by liu-zhengdong**
### Summary / 概述

Fixes #32.

When an upstream extension, middleware, or client injects non-function tools into the request (such as server-side builtin tools like `{ type: "web_search" }`, or custom tools without a `function` wrapper), `tool.function` evaluates to `undefined`. This previously caused an unhandled `TypeError: Cannot read properties of undefined (reading 'parameters')` during schema preparation.

This PR adds defensive null-checks in `slimOpenAIToolsForCursor` and `buildMcpToolDefinitions` to safely handle non-function or malformed tool definitions.

---

### Changes / 变更内容

1. **`slimOpenAIToolsForCursor`**:
   - Safely checks for `tool?.function`.
   - Returns non-function tools unmodified instead of attempting to access `.parameters`.
2. **`buildMcpToolDefinitions`**:
   - Safely skips tool definitions lacking `.function?.name` (since non-function tools cannot be encoded as Cursor MCP client tools).
3. **Tests**:
   - Added unit test in `tests/tool-schema.test.ts` covering non-function tool passing and MCP definition filtering.

---

### Verification / 验证

- Ran `bun run check`: All 276 tests pass, type-check, linter, format, proto, and security checks succeed cleanly.
- Ran `bun run test:coverage`: Coverage requirements satisfied.
- Verified live request handling with mixed tools against Cursor native endpoint.

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved tool preparation to safely handle malformed or incomplete tool definitions without errors.
  - Valid tools continue to be processed with their existing metadata and input schemas.
  - Invalid tools are excluded from generated tool definitions instead of producing unusable output.

### CODE REVIEW

File: src/stream/tool-schema.ts [hunk 1]
```diff
@@ -169,14 +172,16 @@ export function buildMcpToolDefinitions(tools: OpenAIToolDef[]): McpToolDefiniti
...
-      toolName: fn.name,
-      inputSchema,
-    });
-  });
+    result.push(
+      create(McpToolDefinitionSchema, {
+        name: fn.name,
+        description: fn.description || "",
+        providerIdentifier: "pi",
+        toolName: fn.name,
```
[the remark this answers is not in the document]
Comment (liu-zhengdong) [reply src/stream/tool-schema.ts]: Thanks for the detailed review!

1. **Positional indexing:** Callers in the codebase do not rely on 1:1 positional indexing for `mcpTools`. Execution dispatch and steering in `server-messages.ts:availableToolNamesFor` and `native-core.ts` resolve tool calls by `toolName` / `name` rather than array indices. Since non-function or nameless tools cannot be encoded as Cursor MCP client tools anyway, dropping them here is safe and matches the expected wire behavior.
2. **Strict narrowing:** Adopted your suggestion to defensively check that `tool` and `fn` are non-null objects and not arrays (`isObjectRecord`), and verify that `name` is a valid, trimmed non-empty string.
3. **Test coverage:** Extended unit tests with `it.each(["0", "1"])` to test both slim and non-slim modes, including malformed edge cases (`null`, `undefined`, string items, and non-object functions). All 277 tests pass cleanly.
